tests/bsim/bluetooth/mesh/src/test_proxy_sol.c

```diff
@@ -0,0 +1,304 @@
+			FAIL("AppKey add failed (err %d, status %u)", err, status);
+		}
+	}
+
+	err = bt_mesh_cfg_cli_gatt_proxy_set(0, iut_cfg.addr, BT_MESH_GATT_PROXY_DISABLED, &status);
+	if (err || status) {
+		FAIL("Proxy state disable failed (err %d, status %u)", err, status);
+	}
+
+	err = bt_mesh_priv_beacon_cli_gatt_proxy_set(0, iut_cfg.addr, BT_MESH_GATT_PROXY_DISABLED,
+						     &status);
+	if (err || status) {
+		FAIL("Private proxy state disable failed (err %d, status %u)", err, status);
+	}
+
+	ASSERT_OK_MSG(bt_mesh_od_priv_proxy_set(1), "On-Demand Private Proxy enable failed");
+}
+
+static void sol_fixed_pdu_create(struct bt_mesh_subnet *sub, struct net_buf_simple *pdu)
+{
+	static uint32_t fixed_seq_n = 0x123456;
+
+	net_buf_simple_add_u8(pdu, sub->keys[SUBNET_KEY_TX_IDX(sub)].msg.nid);
+	net_buf_simple_add_u8(pdu, 0x80);
+	net_buf_simple_add_le24(pdu, sys_cpu_to_be24(fixed_seq_n));
+	net_buf_simple_add_le16(pdu, sys_cpu_to_be16(bt_mesh_primary_addr()));
+	net_buf_simple_add_le16(pdu, 0x0000);
+
+	ASSERT_OK(bt_mesh_net_encrypt(&sub->keys[SUBNET_KEY_TX_IDX(sub)].msg.enc, pdu, 0,
+				      BT_MESH_NONCE_SOLICITATION));
+
+	ASSERT_OK(bt_mesh_net_obfuscate(pdu->data, 0,
+					&sub->keys[SUBNET_KEY_TX_IDX(sub)].msg.privacy));
+
+	net_buf_simple_push_u8(pdu, 0);
+	net_buf_simple_push_le16(pdu, BT_UUID_MESH_PROXY_SOLICITATION_VAL);
+}
+
+static int sol_fixed_pdu_send(void)
+{
+	NET_BUF_SIMPLE_DEFINE(pdu, 20);
+	net_buf_simple_init(&pdu, 3);
+
+	struct bt_mesh_subnet *sub = bt_mesh_subnet_find(NULL, NULL);
+
+	uint16_t adv_int = BT_MESH_TRANSMIT_INT(CONFIG_BT_MESH_SOL_ADV_XMIT);
+
+	sol_fixed_pdu_create(sub, &pdu);
+
+	struct bt_data ad[] = {
+		BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+		BT_DATA_BYTES(BT_DATA_UUID16_ALL,
+			      BT_UUID_16_ENCODE(BT_UUID_MESH_PROXY_SOLICITATION_VAL)),
+		BT_DATA(BT_DATA_SVC_DATA16, pdu.data, pdu.size),
+	};
+
+	return bt_mesh_adv_bt_data_send(CONFIG_BT_MESH_SOL_ADV_XMIT, adv_int, ad, 3);
+}
+
+static void test_tester_beacon_rcvd(void)
+{
+	tester_configure();
+
+	/* Check that no beacons are currently being picked up by the scanner */
+	ASSERT_EQUAL(-EAGAIN, k_sem_take(&beacon_sem, K_SECONDS(SEM_TIMEOUT)));
+
+	ASSERT_OK(bt_mesh_proxy_solicit(0));
+
+	ASSERT_OK_MSG(k_sem_take(&beacon_sem, K_SECONDS(SEM_TIMEOUT)),
+		      "No beacon (proxy advs) reveiced.");
+
+	PASS();
+}
+
+static void test_tester_immediate_replay_attack(void)
+{
+	tester_configure();
+
+	/* Check that no beacons are currently being picked up by the scanner */
+	ASSERT_EQUAL(-EAGAIN, k_sem_take(&beacon_sem, K_SECONDS(SEM_TIMEOUT)));
+
+	/* Send initial solicitation PDU with fixed sequence number */
+	ASSERT_OK(sol_fixed_pdu_send());
+
+	ASSERT_OK_MSG(k_sem_take(&beacon_sem, K_SECONDS(SEM_TIMEOUT)),
+		      "No beacon (proxy advs) reveiced.");
+
+	/* Wait for iut proxy advs to time out */
+	k_sleep(K_MSEC(200));
+	ASSERT_EQUAL(-EAGAIN, k_sem_take(&beacon_sem, K_SECONDS(SEM_TIMEOUT)));
+
+	k_sem_reset(&beacon_sem);
```

Comment:
This shouldn't be needed as it may affect the next `k_sem_take` if a proxy adv suddenly received in between `k_sem_take` calls.

Reply:
How can I make sure that no beacon from the previous valid solicitation procedure is received during the replay attack phase below if I do not allow the advertisements to time out? Are all proxy advertisements during the same solicitation request identified as the same, so that after the tester has accepted one the following advs will be discarded?

Reply:
The previous `k_sem_take` waits for `-EAGAIN`. So this `k_sem_reset` shouldn't even do anything.

Reply:
Ok. Sorry, I misunderstood which part you wanted me to take out.